### PR TITLE
Nested content optimizations

### PR DESCRIPTION
**Pull request by elit0451**
## Details

<s>- Changing the built-in cache at **NestedContentValues** level for Content Types - before that dictionary cache was never used because it was short-lived and re-created while looping through the items. Now, it is created in **NestedContentPropertyValueEditor** and passed on, so it can be reused.</s>
- Calling `GetValueEditor()` for unique data type ids - reducing the number of objects allocated

### CODE REVIEW

**Review by nikolajlauridsen — APPROVED** [review]
Looks good to me 👍 💪

**Review by mikecp — APPROVED** [review]

**Comment by mikecp** [comment]
Thanks @elit0451 for this update!
Unfortunately I can't merge it as it is a branch on the umbraco/Umbraco-CMS repo directly...

File: src/Umbraco.Web/PropertyEditors/NestedContentPropertyEditor.cs [hunk 1]
```diff
@@ -48,6 +49,7 @@ public NestedContentPropertyEditor(ILogger logger, Lazy<PropertyEditorCollection
             _dataTypeService = dataTypeService;
             _contentTypeService = contentTypeService;
             _localizedTextService = localizedTextService;
+            _contentTypes = new Lazy<Dictionary<string, IContentType>>(() => contentTypeService.GetAll().ToDictionary(c => c.Alias));
```
Comment (Shazwazza) [inline src/Umbraco.Web/PropertyEditors/NestedContentPropertyEditor.cs:52]: Sorry, same issue as I mention in the block list change https://github.com/umbraco/Umbraco-CMS/pull/10235#pullrequestreview-657110265, this part will need to be reverted.

File: src/Umbraco.Web/PropertyEditors/NestedContentPropertyEditor.cs [hunk 2]
```diff
@@ -184,9 +187,16 @@ public override object ToEditor(Property property, IDataTypeService dataTypeServ
                                 continue;
                             }
 
-                            var tempConfig = dataTypeService.GetDataType(prop.Value.PropertyType.DataTypeId).Configuration;
-                            var valEditor = propEditor.GetValueEditor(tempConfig);
-                            var convValue = valEditor.ToEditor(tempProp, dataTypeService);
+                            var dataTypeId = prop.Value.PropertyType.DataTypeId;
+                            if (!valEditors.ContainsKey(dataTypeId))
```
Comment (Shazwazza) [inline src/Umbraco.Web/PropertyEditors/NestedContentPropertyEditor.cs:191]: valEditors.TryGetValue is a faster way to lookup and retrieve from a dictionary at the same time.

**Review by Shazwazza — CHANGES_REQUESTED** [review]
Left some notes inline

**Comment by skttl** [comment]
Is it possible that the optimization in `ToEditor()` can be applied to `FromEditor()` too?

Doc Type Grid Editor suffers from sql calls in FromEditor - https://github.com/skttl/umbraco-doc-type-grid-editor/issues/194

File: src/Umbraco.Web/PropertyEditors/NestedContentPropertyEditor.cs (hunk 1 above)
Comment (elit0451) [reply src/Umbraco.Web/PropertyEditors/NestedContentPropertyEditor.cs]: Addressed

**Comment by nul800sebastiaan** [comment]
Cherry picked for 8.15.0 3e89e468406f4eb3ad33522bcad8a357fc7bff14

**Pull request merged**